### PR TITLE
Akeneo Create a new Product Media file

### PR DESCRIPTION
**Pull request by feyzullah**
Could not make it work. There should be something wrong with `FormData` but I could not figure out yet. Opening as a draft PR.

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/akeneo/actions/create-a-new-product-media-file/create-a-new-product-media-file.mjs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      throw new ConfigurationError("Either `Product Identifier` or `Product Model Code` should be set!");
+    }
+    const path = utils.checkTmp(this.filename);
+    if (!fs.existsSync(path)) {
+      throw new ConfigurationError("File does not exist!");
+    }
+    const payload = {
+      attribute: this.mediaFileAttributeCode,
+      scope: null,
+      locale: null,
```
Comment (andrewjschuang) [inline components/akeneo/actions/create-a-new-product-media-file/create-a-new-product-media-file.mjs:50]: Are these keys necessary?

File: components/akeneo/actions/create-a-new-product-media-file/create-a-new-product-media-file.mjs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    const fileParts = path.split("/");
+    data.append("file", file, fileParts[fileParts.length - 1]);
+    const contentLength = await new Promise((resolve, reject) => {
+      data.getLength((err, length) => {
+        if (err) {
+          reject(err);
+        }
+        resolve(length);
+      });
+    });
```
Comment (andrewjschuang) [inline components/akeneo/actions/create-a-new-product-media-file/create-a-new-product-media-file.mjs:70]:
```suggestion
    const contentLength = data.getLengthSync();
```
Can we use this function?

File: components/akeneo/akeneo.app.mjs [hunk 3]
```diff
@@ -1,11 +1,168 @@
+import { axios } from "@pipedream/platform";
+
+const pageSize = 25;
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:3]:
```suggestion
const PAGE_SIZE = 25;
```
Convention for constants

File: components/akeneo/akeneo.app.mjs [hunk 4]
```diff
@@ -1,11 +1,168 @@
...
+
 export default {
   type: "app",
   app: "akeneo",
-  propDefinitions: {},
+  propDefinitions: {
+    productId: {
+      type: "string",
+      label: "Product Identifier",
+      description: "An ID identifying the product. - Either this prop or `Product Model Code` should be set.",
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:12]:
```suggestion
      description: "An ID identifying the product - either this prop or `Product Model Code` should be set.",
```
Nitpick: either use ` - ` or just the period.

File: components/akeneo/akeneo.app.mjs [hunk 5]
```diff
@@ -1,11 +1,168 @@
...
+        return resp?._embedded?.items?.map((product) => ({
+          label: product.uuid,
+          value: `${product.family} - ${product.identifier}`,
+        }));
+      },
+    },
+    productModelCode: {
+      type: "string",
+      label: "Product Model Code",
+      description: "A code identifying the product model. - Either this prop or `Product Identifier` should be set.",
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:31]: Nitpick: either use ` - ` or just the period.

File: components/akeneo/akeneo.app.mjs [hunk 6]
```diff
@@ -1,11 +1,168 @@
...
+        const resp = await this.getProducts({
+          params: {
+            limit: pageSize,
+            page,
+          },
+        });
+        return resp?._embedded?.items?.map((product) => ({
+          label: product.uuid,
+          value: `${product.family} - ${product.identifier}`,
+        }));
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:25]:
```suggestion
        })) || [];
```
Better to return an empty array as a fail-safe protocol if the optional chaining returns `undefined`

File: components/akeneo/akeneo.app.mjs [hunk 7]
```diff
@@ -1,11 +1,168 @@
...
+      optional: true,
+      async options({ page }) {
+        page++;
+        const resp = await this.getProductModels({
+          params: {
+            limit: pageSize,
+            page,
+          },
+        });
+        return resp?._embedded?.items?.map((model) => model.code);
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:41]:
```suggestion
        return resp?._embedded?.items?.map((model) => model.code) || [];
```

File: components/akeneo/akeneo.app.mjs [hunk 8]
```diff
@@ -1,11 +1,168 @@
...
+                  operator: "IN",
+                  value: [
+                    "pim_catalog_image",
+                  ],
+                },
+              ],
+            },
+          },
+        });
+        return resp?._embedded?.items?.map((attribute) => attribute.code);
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:66]:
```suggestion
        return resp?._embedded?.items?.map((attribute) => attribute.code) || [];
```

File: components/akeneo/akeneo.app.mjs [hunk 9]
```diff
@@ -1,11 +1,168 @@
...
+      description: "A code identifying the channel",
+      async options({ page }) {
+        page++;
+        const resp = await this.getChannels({
+          params: {
+            limit: pageSize,
+            page,
+          },
+        });
+        return resp?._embedded?.items?.map((channel) => channel.code);
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:81]:
```suggestion
        return resp?._embedded?.items?.map((channel) => channel.code) || [];
```

File: components/akeneo/akeneo.app.mjs [hunk 10]
```diff
@@ -1,11 +1,168 @@
...
+    },
+    localeCode: {
+      type: "string",
+      label: "Locale Code",
+      description: "A code identifying the locale",
+      async options({ channelCode }) {
+        const resp = await this.getChannel({
+          channelCode,
+        });
+        return resp?.locales;
```
Comment (andrewjschuang) [inline components/akeneo/akeneo.app.mjs:92]:
```suggestion
        return resp?.locales || [];
```

File: components/akeneo/actions/create-a-new-product-media-file/create-a-new-product-media-file.mjs (hunk 1 above)
Comment (feyzullah) [reply components/akeneo/actions/create-a-new-product-media-file/create-a-new-product-media-file.mjs:50]: yes, they are

**Comment by feyzullah** [comment]
Hi @andrewjschuang, I applied all your suggestions except the one about null values. That object is serialized as JSON and the API dictates its format. Because of this, those null values are necessary.
Can you please check this PR once more?

**Review by andrewjschuang — APPROVED** [review]
LGTM!

**Comment by DilanAthukorala** [comment]
Verification completed, ready for release!

**Pull request merged**